specification/dartLangSpec.tex

```diff
@@ -9685,29 +9691,113 @@ \subsubsection{Cascaded Invocations}
+The motivation for having a cascaded member access is that it allows for
+performing a chain of operations based on an object
+while preserving a reference to that object for further processing.%
+}
+
+\commentary{%
+For example, \code{C()..foo.bar = 2} allows us to obtain a reference to
+the object $o$ which is the result of evaluating \code{C()},
+and at the same time use $o$ to invoke the getter \code{foo} and
+the setter \code{bar=} on the value returned by that getter.%
+}
 
 \LMHash{}%
-The static type of $c$ is the static type of $e$.
+An expression of the form \code{$e_0$?..$s$} where
+$e_0$ is a \synt{conditionalExpression} and
+$s$ is a \synt{cascadeSection} is an
+\IndexCustom{initially conditional cascaded member access}{%
+  cascaded member access!initially conditional}.
+Moreover, if \code{$e_0$} is
+an initially conditional cascaded member access and
+$s$ is derived from \synt{cascadeSection} then
+\code{$e_0$..$s$} is also initially conditional.
 
-\rationale{
-With the introduction of null-aware conditional assignable expressions (\ref{assignableExpressions}), it would make sense to extend cascades with a null-aware conditional form as well.
-One might define \code{$e$?..\metavar{suffix}} to be equivalent to the expression \code{$t$ == \NULL{} ? \NULL{} : $t$.\metavar{suffix}} where $t$ is a fresh variable bound to the value of $e$.
+\commentary{%
+In short, a cascade is initially conditional if
+the ``innermost dots'' are \lit{?..} rather than \lit{..}.
+Note that only the innermost dots can have the \lit{?}.
+All the non-innermost ones are implicitly skipped if the receiver is null,
+so any \lit{?} on a non-innermost \lit{..} would be useless.%
+}
+
+\LMHash{}%
+Let $s$ be a \synt{cascadeSection}.
+Let $e$ be an expression of the form \code{$e_0$..$s$}
+where $e_0$ is an \synt{expression}
+such that $e$ is not initially conditional.
+Exactly the same compile-time errors are raised for $e$
+as would be raised for \code{$e_0$.$s$}.
+The static type of $e$ is the static type of $e_0$.
+
+\LMHash{}%
+Let $s$ be a \synt{cascadeSection}.
+If $e$ is an expression of the form \code{$e_0$?..$s$}
+where $e_0$ is a \synt{conditionalExpression},
+exactly the same compile-time errors are raised for $e$
+% This rule is NNBD-ready, so we use it even though we could
+% use a slightly simpler set of rules today.
+as would be raised for \code{$e_0$?.$s$}.
+If $e$ is an expression of the form \code{$e_0$..$s$}
+where $e_0$ is an \synt{expression}
+such that $e$ is initially conditional,
+exactly the same compile-time errors are raised for $e$
+as would be raised for \code{$e_0$?.$s$}.
+In both cases, the static type of $e$ is the static type of $e_0$.
+
+\LMHash{}%
+Evaluation of a cascaded member access $e$
+of the form \code{$e_0$..$s$}
+where $e$ is an \synt{expression} and
+$s$ is a \synt{cascadeSection}
+such that $e$ is not initially conditional
+proceeds as follows:
+Evaluate $e_0$ to an object $o$.
```

Comment:
Could you just insert here:

If $e_0$ is initially conditional and $o$ is the null object, then $e$ evaluates to the null object.
Otherwise let $t$ .... 

and drop the "is not initially conditional" clause above? Then the entire next section should be subsumed by that.

Reply:
Please follow up if needed&mdash;I have not made this change.

The reason for not doing this is that the following section spells out the shorting semantics, and if I make this change then we won't describe that semantics, we will describe the approach where we step through each layer of the cascade one section at a time.

Didn't we want to make the shorting semantics explicit?